### PR TITLE
Choose random free ports atomically

### PR DESCRIPTION
**Pull request by f4lco**
Now Gretty should choose "random free ports" atomically for both: the service protocol and the server connectors. This replaces the prior approach of "port = bind(0); unbind(port); ...; bind(port)", which lets two concurrent Gretty invocations choose identical ports.

Implementation:
* the service protocol server sockets must immediately bind and use port zero. On the "service" side (subprocess containing the server connector) the service protocol changes slightly to include the "service port" in the "init" message.
* The server connectors must also bind to port zero.
* In both instances, whenever information is required about the actually bound port, the caller has to query the "localPort" property. "port" is still zero, and indicates that the port was not assigned a priori.

Other benefits:

* Dropped two implementations of "choose random free port"
* Deduped two implementations of the "service protocol"

Concerns:
* API stability: externally, Gretty still accepts "-1" as sentinel value for "choose random free port". I think this is alright but still has some potential for bugs, if the translation from config value "-1" to bound port "0" was missing.
* API stability: this patch removes already deprecated port properties (`servicePort`, `statusPort`) 
 from tasks.
* API stability: the task property `auxPortRange` is removed. I do not think it can be meaningfully supported.
* Exhaustiveness: Gretty must use the `localPort` property everywhere.
* Breaking things I do not use directly: this de-flaked version of Gretty works well for integration tests at my current organization. However, that deployment uses only a subset of Gretty's features, and uses only the Tomcat connector. There appear to be plenty of tests, but I am unsure about coverage of some of the less popular features / tasks.

### CODE REVIEW

**Comment by f4lco** [comment]
I just spotted an irregular log message:

> 11:08:56.972 [main] DEBUG org.akhikhl.gretty.ServiceProtocol - ServiceProtocol.readMessageFromServerSocket(42541) -> {"servletContainerId":"tomcat9","servletContainerDescription":"Tomcat 9.0.13","httpEnabled":true,**"httpPort":-1**,"loggingLevel":"INFO","consoleLogEnabled":true," <snip>

It works, but I think that the actually chosen port should be reflected, and not the config value of "-1".

- [ ] check

**Comment by javabrett** [comment]
Thanks for the PR. Reviewing.

File: libs/gretty-common/src/main/groovy/org/akhikhl/gretty/ServiceProtocol.groovy [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+ */
+package org.akhikhl.gretty
+
+import groovy.util.logging.Slf4j
+
+import java.nio.charset.StandardCharsets
+import java.util.concurrent.CompletableFuture
+
+@Slf4j
+final class ServiceProtocol {
```
[the remark this answers is not in the document]
Comment (javabrett) [reply libs/gretty-common/src/main/groovy/org/akhikhl/gretty/ServiceProtocol.groovy]: NVM I can see that there's currently no shared classpath there ...

**Comment by f4lco** [comment]
Thanks for reviewing, @javabrett. I checked the log: the message is indeed expected. This makes this PR complete / ready-to-merge?

**Comment by f4lco** [comment]
It's been a couple of weeks. @javabrett any update on your side? Are there any open to-dos?

**Comment by javabrett** [comment]
Thanks @f4lco for this contribution - it has been released in Gretty 3.0.2.

There is one follow-up item - there's some logging that isn't properly thresholded by-default, at `DEBUG` level - might need to check if we have a package logger for this, or for logback configuration in the context it is running in.  See: https://travis-ci.org/github/gretty-gradle-plugin/gretty/jobs/668574695#L921

Starts with:

```
04:04:36.760 [main] DEBUG org.akhikhl.gretty.ServiceProtocol - ServiceProtocol.readMessageFromServerSocket(37399) -> {"servletContainerId":"tomcat9","servletContainerDescription":"Tomcat 9.0.31","httpEnabled":true,"loggingLevel":"INFO","consoleLogEnabled":true,"fileLogEnabled":true,"logFileName":"extraResourceBases","logDir":"/home/travis/logs","baseDir":"/home/travis/build/gretty-gradle-plugin/gretty/integrationTests/extraResourceBases/build/serverBaseDir_tomcat9","webApps":
```

**Comment by boris-petrov** [comment]
@f4lco - #147 also seems to be caused by the `ServiceProtocol` class. The offending line is `ServiceProtocol.groovy:76` which is `Socket s = new Socket(InetAddress.localHost, port)`. Not sure what exactly is going on. Would you mind taking a look please?

**Comment by f4lco** [comment]
@javabrett thanks! Sure, I opened #150 and put up an proposal for discussion.

@boris-petrov No problem, I opened https://github.com/gretty-gradle-plugin/gretty/pull/149 with a potential fix.

**Pull request merged**